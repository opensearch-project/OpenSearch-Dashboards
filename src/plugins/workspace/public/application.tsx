/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React from 'react';
import ReactDOM from 'react-dom';
import { AppMountParameters, ScopedHistory } from '../../../core/public';
import { OpenSearchDashboardsContextProvider } from '../../opensearch_dashboards_react/public';
import { WorkspaceFatalError } from './components/workspace_fatal_error';
import { WorkspaceCreatorApp } from './components/workspace_creator_app';
import { WorkspaceUpdaterApp } from './components/workspace_updater_app';
import { WorkspaceListApp } from './components/workspace_list_app';
import { Services } from './types';

export const renderCreatorApp = ({ element }: AppMountParameters, services: Services) => {
  ReactDOM.render(
    <OpenSearchDashboardsContextProvider services={services}>
      <WorkspaceCreatorApp />
    </OpenSearchDashboardsContextProvider>,
    element
  );

  return () => {
    ReactDOM.unmountComponentAtNode(element);
  };
};

export const renderUpdaterApp = ({ element }: AppMountParameters, services: Services) => {
  ReactDOM.render(
    <OpenSearchDashboardsContextProvider services={services}>
      <WorkspaceUpdaterApp />
    </OpenSearchDashboardsContextProvider>,
    element
  );

  return () => {
    ReactDOM.unmountComponentAtNode(element);
  };
};

export const renderFatalErrorApp = (params: AppMountParameters, services: Services) => {
  const { element } = params;
  const history = params.history as ScopedHistory<{ error?: string }>;
  ReactDOM.render(
    <OpenSearchDashboardsContextProvider services={services}>
      <WorkspaceFatalError error={history.location.state?.error} />
    </OpenSearchDashboardsContextProvider>,
    element
  );

  return () => {
    ReactDOM.unmountComponentAtNode(element);
  };
};
export const renderListApp = ({ element }: AppMountParameters, services: Services) => {
  ReactDOM.render(
    <OpenSearchDashboardsContextProvider services={services}>
      <WorkspaceListApp />
    </OpenSearchDashboardsContextProvider>,
    element
  );

  return () => {
    ReactDOM.unmountComponentAtNode(element);
  };
};
