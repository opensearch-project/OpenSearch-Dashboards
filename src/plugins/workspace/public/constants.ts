/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { i18n } from '@osd/i18n';

import { WorkspaceCollaboratorAccessLevel } from './types';
import { WorkspacePermissionMode } from '../common/constants';

export const WORKSPACE_ACCESS_LEVEL_NAMES: { [key in WorkspaceCollaboratorAccessLevel]: string } = {
  readOnly: i18n.translate('workspace.accessLevel.readOnlyName', {
    defaultMessage: 'Read only',
  }),
  readAndWrite: i18n.translate('workspace.accessLevel.readAndWriteName', {
    defaultMessage: 'Read and write',
  }),
  admin: i18n.translate('workspace.accessLevel.AdminName', {
    defaultMessage: 'Admin',
  }),
};

export const accessLevelNameToWorkspacePermissionModesMap: {
  [key in WorkspaceCollaboratorAccessLevel]: WorkspacePermissionMode[];
} = {
  readOnly: [WorkspacePermissionMode.LibraryRead, WorkspacePermissionMode.Read],
  readAndWrite: [WorkspacePermissionMode.LibraryWrite, WorkspacePermissionMode.Read],
  admin: [WorkspacePermissionMode.LibraryWrite, WorkspacePermissionMode.Write],
};
