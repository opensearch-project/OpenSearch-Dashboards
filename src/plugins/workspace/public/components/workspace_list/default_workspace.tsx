/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React from 'react';
import { BehaviorSubject, of } from 'rxjs';
import { useObservable } from 'react-use';
import { EuiFlexItem, EuiPanel, EuiFlexGroup, EuiText, EuiHorizontalRule } from '@elastic/eui';
import { OpenSearchDashboardsContextProvider } from '../../../../opensearch_dashboards_react/public';
import { WorkspaceListInner } from '.';
import { Services, WorkspaceUseCase } from '../../types';

interface Props {
  services: Services;
  registeredUseCases$: BehaviorSubject<WorkspaceUseCase[]>;
}

export const UserDefaultWorkspace = ({ services, registeredUseCases$ }: Props) => {
  const { workspaces } = services;
  const workspaceList = useObservable(workspaces?.workspaceList$ ?? of([]), []);

  return (
    <OpenSearchDashboardsContextProvider services={services}>
      <EuiPanel>
        <EuiFlexGroup direction="column" gutterSize="xs">
          <EuiFlexItem>
            <EuiText>
              <h3>Workspaces ({workspaceList?.length})</h3>
            </EuiText>
          </EuiFlexItem>
          <EuiFlexItem>
            <EuiHorizontalRule margin="s" />
          </EuiFlexItem>
          <EuiFlexItem>
            <WorkspaceListInner
              fullPage={false}
              selectable={false}
              searchable={false}
              registeredUseCases$={registeredUseCases$}
              excludedActionNames={['edit', 'delete']}
              includedColumns={[
                { field: 'name', width: '25%' },
                { field: 'useCase', width: '20%' },
                { field: 'description', width: '25%' },
                { field: 'permissionMode', width: '10%' },
              ]}
            />
          </EuiFlexItem>
        </EuiFlexGroup>
      </EuiPanel>
    </OpenSearchDashboardsContextProvider>
  );
};
