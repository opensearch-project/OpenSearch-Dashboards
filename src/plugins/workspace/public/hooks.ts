/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { ApplicationStart, PublicAppInfo } from 'opensearch-dashboards/public';
import { useObservable } from 'react-use';
import { useMemo } from 'react';

export function useApplications(applicationInstance: ApplicationStart) {
  const applications = useObservable(applicationInstance.applications$);
  return useMemo(() => {
    const apps: PublicAppInfo[] = [];
    applications?.forEach((app) => {
      apps.push(app);
    });
    return apps;
  }, [applications]);
}
