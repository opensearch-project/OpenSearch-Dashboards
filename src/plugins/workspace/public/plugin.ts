/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { Plugin } from '../../../core/public';

export class WorkspacePlugin implements Plugin<{}, {}, {}> {
  public async setup() {
    return {};
  }

  public start() {
    return {};
  }

  public stop() {}
}
