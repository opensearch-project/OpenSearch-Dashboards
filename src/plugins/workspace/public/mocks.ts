/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { BehaviorSubject } from 'rxjs';
import { WORKSPACE_USE_CASES } from '../common/constants';

export const createMockedRegisteredUseCases = () =>
  [
    WORKSPACE_USE_CASES.observability,
    WORKSPACE_USE_CASES['security-analytics'],
    WORKSPACE_USE_CASES.essentials,
    WORKSPACE_USE_CASES.search,
  ].map((item) => ({
    ...item,
    features: item.features.map((id) => ({ id })),
  }));

export const createMockedRegisteredUseCases$ = () =>
  new BehaviorSubject(createMockedRegisteredUseCases());
