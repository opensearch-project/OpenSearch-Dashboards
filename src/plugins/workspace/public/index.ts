/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { WorkspacePlugin } from './plugin';

export function plugin() {
  return new WorkspacePlugin();
}
