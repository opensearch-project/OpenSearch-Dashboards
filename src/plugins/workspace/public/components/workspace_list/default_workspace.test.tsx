/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React from 'react';
import { BehaviorSubject } from 'rxjs';
import { render } from '@testing-library/react';
import { I18nProvider } from '@osd/i18n/react';
import { coreMock } from '../../../../../core/public/mocks';
import { createMockedRegisteredUseCases$ } from '../../mocks';
import { UserDefaultWorkspace } from './default_workspace';
import { WorkspaceClient } from '../../workspace_client';
import { WorkspaceObject } from '../../../../../core/public';

jest.mock('../utils/workspace');
jest.mock('../../utils', () => {
  const original = jest.requireActual('../../utils');
  return {
    ...original,
    getDataSourcesList: jest.fn().mockResolvedValue(() => [
      {
        id: 'ds_id1',
        title: 'ds_title1',
        workspaces: 'id1',
      },
      {
        id: 'ds_id2',
        title: 'ds_title2',
        workspaces: 'id1',
      },
      {
        id: 'ds_id3',
        title: 'ds_title3',
        workspaces: 'id1',
      },
    ]),
  };
});

function getWrapUserDefaultWorkspaceList(
  workspaceList = [
    {
      id: 'id1',
      name: 'name1',
      features: ['use-case-all'],
      description:
        'should be able to see the description tooltip when hovering over the description',
      lastUpdatedTime: '1999-08-06T02:00:00.00Z',
      permissions: {
        write: {
          users: ['admin', 'nonadmin'],
        },
      },
    },
    {
      id: 'id2',
      name: 'name2',
      features: ['use-case-observability'],
      description:
        'should be able to see the description tooltip when hovering over the description',
      lastUpdatedTime: '1999-08-06T00:00:00.00Z',
    },
    {
      id: 'id3',
      name: 'name3',
      features: ['use-case-search'],
      description: '',
      lastUpdatedTime: '1999-08-06T01:00:00.00Z',
    },
  ],
  isDashboardAdmin = true
) {
  const coreStartMock = coreMock.createStart();
  const coreSetupMock = coreMock.createSetup();
  coreStartMock.application.capabilities = {
    ...coreStartMock.application.capabilities,
    dashboards: {
      isDashboardAdmin,
    },
  };

  const mockHeaderControl = ({ controls }) => {
    return controls?.[0].description ?? controls?.[0].renderComponent ?? null;
  };

  const workspaceClientMock = new WorkspaceClient(coreSetupMock.http, coreSetupMock.workspaces);

  const services = {
    ...coreStartMock,
    workspaces: {
      ...coreStartMock.workspaces,
      workspaceList$: new BehaviorSubject<WorkspaceObject[]>(workspaceList),
    },
    uiSettings: {
      ...coreStartMock.uiSettings,
      get: jest.fn().mockImplementation((key) => {
        if (key === 'dateFormat') {
          return 'MMM D, YYYY @ HH:mm:ss.SSS';
        }
        return null;
      }),
    },
    navigationUI: {
      HeaderControl: mockHeaderControl,
    },
    workspaceClient: workspaceClientMock,
  };

  return (
    <I18nProvider>
      <UserDefaultWorkspace
        registeredUseCases$={createMockedRegisteredUseCases$()}
        services={services}
      />
    </I18nProvider>
  );
}

describe('UserDefaultWorkspace', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should render title and table normally', () => {
    const { getByText, getByRole, container } = render(getWrapUserDefaultWorkspaceList());
    expect(getByText('Workspaces (3)')).toBeInTheDocument();
    expect(getByRole('table')).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  it('should render data in table based on workspace list data', async () => {
    const { getByText, queryByText, queryAllByText } = render(getWrapUserDefaultWorkspaceList());

    // should display workspace names
    expect(getByText('name1')).toBeInTheDocument();
    expect(getByText('name2')).toBeInTheDocument();

    // should display use case
    expect(getByText('Analytics (All)')).toBeInTheDocument();
    expect(getByText('Observability')).toBeInTheDocument();

    // owner column not display
    expect(queryByText('admin')).not.toBeInTheDocument();

    // euiTableRow-isSelectable
    expect(document.querySelectorAll('.euiTableRow-isSelectable').length).toBe(0);

    // action button Set as default in document
    expect(queryAllByText('Set as my default')).toHaveLength(3);
  });
});
