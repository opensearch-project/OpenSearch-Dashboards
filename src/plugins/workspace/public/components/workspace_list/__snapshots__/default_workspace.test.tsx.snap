// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserDefaultWorkspace should render title and table normally 1`] = `
<div>
  <div
    class="euiPanel euiPanel--paddingMedium euiPanel--borderRadiusMedium euiPanel--plain euiPanel--hasShadow"
  >
    <div
      class="euiFlexGroup euiFlexGroup--gutterExtraSmall euiFlexGroup--directionColumn euiFlexGroup--responsive"
    >
      <div
        class="euiFlexItem"
      >
        <div
          class="euiText euiText--medium"
        >
          <h3>
            Workspaces (
            3
            )
          </h3>
        </div>
      </div>
      <div
        class="euiFlexItem"
      >
        <hr
          class="euiHorizontalRule euiHorizontalRule--full euiHorizontalRule--marginSmall"
        />
      </div>
      <div
        class="euiFlexItem"
      >
        <div
          class="euiBasicTable"
        >
          <div>
            <div
              class="euiTableHeaderMobile"
            >
              <div
                class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsBaseline euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow"
              >
                <div
                  class="euiFlexItem euiFlexItem--flexGrowZero"
                />
                <div
                  class="euiFlexItem euiFlexItem--flexGrowZero"
                >
                  <div
                    class="euiTableSortMobile"
                  >
                    <div
                      class="euiPopover euiPopover--anchorDownRight"
                    >
                      <div
                        class="euiPopover__anchor"
                      >
                        <button
                          class="euiButtonEmpty euiButtonEmpty--primary euiButtonEmpty--xSmall euiButtonEmpty--flushRight"
                          type="button"
                        >
                          <span
                            class="euiButtonContent euiButtonContent--iconRight euiButtonEmpty__content"
                          >
                            <span
                              class="euiButtonContent__icon"
                              color="inherit"
                              data-euiicon-type="arrowDown"
                            />
                            <span
                              class="euiButtonEmpty__text"
                            >
                              Sorting
                            </span>
                          </span>
                        </button>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <table
              class="euiTable euiTable--compressed euiTable--responsive"
              id="__table_generated-id"
              tabindex="-1"
            >
              <caption
                class="euiScreenReaderOnly euiTableCaption"
              />
              <thead>
                <tr>
                  <th
                    aria-live="polite"
                    aria-sort="ascending"
                    class="euiTableHeaderCell"
                    data-test-subj="tableHeaderCell_name_0"
                    role="columnheader"
                    scope="col"
                    style="width: 25%;"
                  >
                    <button
                      class="euiTableHeaderButton euiTableHeaderButton-isSorted"
                      data-test-subj="tableHeaderSortButton"
                      type="button"
                    >
                      <span
                        class="euiTableCellContent"
                      >
                        <span
                          class="euiTableCellContent__text"
                          title="Name"
                        >
                          Name
                        </span>
                        <span
                          class="euiTableSortIcon"
                          data-euiicon-type="sortUp"
                        />
                      </span>
                    </button>
                  </th>
                  <th
                    class="euiTableHeaderCell"
                    data-test-subj="tableHeaderCell_useCase_1"
                    role="columnheader"
                    scope="col"
                    style="width: 20%;"
                  >
                    <span
                      class="euiTableCellContent"
                    >
                      <span
                        class="euiTableCellContent__text"
                        title="Use case"
                      >
                        Use case
                      </span>
                    </span>
                  </th>
                  <th
                    class="euiTableHeaderCell"
                    data-test-subj="tableHeaderCell_description_2"
                    role="columnheader"
                    scope="col"
                    style="width: 25%;"
                  >
                    <span
                      class="euiTableCellContent"
                    >
                      <span
                        class="euiTableCellContent__text"
                        title="Description"
                      >
                        Description
                      </span>
                    </span>
                  </th>
                  <th
                    class="euiTableHeaderCell"
                    data-test-subj="tableHeaderCell_permissionMode_3"
                    role="columnheader"
                    scope="col"
                    style="width: 10%;"
                  >
                    <span
                      class="euiTableCellContent"
                    >
                      <span
                        class="euiTableCellContent__text"
                        title="Permissions"
                      >
                        Permissions
                      </span>
                    </span>
                  </th>
                  <th
                    class="euiTableHeaderCell"
                    role="columnheader"
                    scope="col"
                  >
                    <span
                      class="euiTableCellContent euiTableCellContent--alignRight"
                    >
                      <span
                        class="euiTableCellContent__text"
                        title="Actions"
                      >
                        Actions
                      </span>
                    </span>
                  </th>
                </tr>
              </thead>
              <tbody>
                <tr
                  class="euiTableRow euiTableRow-hasActions"
                >
                  <td
                    class="euiTableRowCell"
                    style="width: 25%;"
                  >
                    <div
                      class="euiTableRowCell__mobileHeader euiTableRowCell--hideForDesktop"
                    >
                      Name
                    </div>
                    <div
                      class="euiTableCellContent euiTableCellContent--overflowingContent"
                    >
                      <span
                        class=""
                      >
                        <a
                          class="euiLink euiLink--primary"
                          href="http://localhost/w/id1/undefined"
                          rel="noreferrer"
                        >
                          <div
                            class="euiFlexGroup euiFlexGroup--gutterExtraSmall euiFlexGroup--alignItemsCenter euiFlexGroup--directionRow euiFlexGroup--responsive"
                          >
                            <div
                              class="euiFlexItem"
                            >
                              <div
                                class="euiText euiText--small"
                              >
                                name1
                              </div>
                            </div>
                          </div>
                        </a>
                      </span>
                    </div>
                  </td>
                  <td
                    class="euiTableRowCell"
                    style="width: 20%;"
                  >
                    <div
                      class="euiTableRowCell__mobileHeader euiTableRowCell--hideForDesktop"
                    >
                      Use case
                    </div>
                    <div
                      class="euiTableCellContent"
                    >
                      <span
                        class="euiTableCellContent__text"
                      >
                        Analytics
                      </span>
                    </div>
                  </td>
                  <td
                    class="euiTableRowCell"
                    style="width: 25%;"
                  >
                    <div
                      class="euiTableRowCell__mobileHeader euiTableRowCell--hideForDesktop"
                    >
                      Description
                    </div>
                    <div
                      class="euiTableCellContent euiTableCellContent--overflowingContent"
                    >
                      <span
                        class="euiToolTipAnchor"
                      >
                        <div
                          class="euiText euiText--extraSmall eui-textTruncate"
                          style="max-width: 150px;"
                        >
                          should be able to see the description tooltip when hovering over the description
                        </div>
                      </span>
                    </div>
                  </td>
                  <td
                    class="euiTableRowCell"
                    style="width: 10%;"
                  >
                    <div
                      class="euiTableRowCell__mobileHeader euiTableRowCell--hideForDesktop"
                    >
                      Permissions
                    </div>
                    <div
                      class="euiTableCellContent euiTableCellContent--overflowingContent"
                    >
                      <span
                        class="euiToolTipAnchor"
                      >
                        <div
                          class="euiText euiText--extraSmall"
                        >
                          Admin
                        </div>
                      </span>
                    </div>
                  </td>
                  <td
                    class="euiTableRowCell euiTableRowCell--hasActions"
                  >
                    <div
                      class="euiTableCellContent euiTableCellContent--alignRight euiTableCellContent--showOnHover euiTableCellContent--overflowingContent"
                    >
                      <span
                        class="euiToolTipAnchor"
                      >
                        <button
                          aria-labelledby="generated-id"
                          class="euiButtonIcon euiButtonIcon--primary euiButtonIcon--empty euiButtonIcon--xSmall euiTableCellContent__hoverItem"
                          data-test-subj="workspace-list-copy-id-icon"
                          type="button"
                        >
                          <span
                            aria-hidden="true"
                            class="euiButtonIcon__icon"
                            color="inherit"
                            data-euiicon-type="copy"
                          />
                        </button>
                        <span
                          class="euiScreenReaderOnly"
                          id="generated-id"
                        >
                          <div
                            class="euiText euiText--medium"
                          >
                            Copy ID
                          </div>
                        </span>
                      </span>
                      <span
                        class="euiToolTipAnchor"
                      >
                        <button
                          aria-labelledby="generated-id"
                          class="euiButtonIcon euiButtonIcon--primary euiButtonIcon--empty euiButtonIcon--xSmall euiTableCellContent__hoverItem"
                          data-test-subj="workspace-list-set-default-icon"
                          type="button"
                        >
                          <span
                            aria-hidden="true"
                            class="euiButtonIcon__icon"
                            color="inherit"
                            data-euiicon-type="flag"
                          />
                        </button>
                        <span
                          class="euiScreenReaderOnly"
                          id="generated-id"
                        >
                          <div
                            class="euiText euiText--medium"
                          >
                            Set as my default
                          </div>
                        </span>
                      </span>
                    </div>
                  </td>
                </tr>
                <tr
                  class="euiTableRow euiTableRow-hasActions"
                >
                  <td
                    class="euiTableRowCell"
                    style="width: 25%;"
                  >
                    <div
                      class="euiTableRowCell__mobileHeader euiTableRowCell--hideForDesktop"
                    >
                      Name
                    </div>
                    <div
                      class="euiTableCellContent euiTableCellContent--overflowingContent"
                    >
                      <span
                        class=""
                      >
                        <a
                          class="euiLink euiLink--primary"
                          href="http://localhost/w/id2/undefined"
                          rel="noreferrer"
                        >
                          <div
                            class="euiFlexGroup euiFlexGroup--gutterExtraSmall euiFlexGroup--alignItemsCenter euiFlexGroup--directionRow euiFlexGroup--responsive"
                          >
                            <div
                              class="euiFlexItem"
                            >
                              <div
                                class="euiText euiText--small"
                              >
                                name2
                              </div>
                            </div>
                          </div>
                        </a>
                      </span>
                    </div>
                  </td>
                  <td
                    class="euiTableRowCell"
                    style="width: 20%;"
                  >
                    <div
                      class="euiTableRowCell__mobileHeader euiTableRowCell--hideForDesktop"
                    >
                      Use case
                    </div>
                    <div
                      class="euiTableCellContent"
                    >
                      <span
                        class="euiTableCellContent__text"
                      >
                        Observability
                      </span>
                    </div>
                  </td>
                  <td
                    class="euiTableRowCell"
                    style="width: 25%;"
                  >
                    <div
                      class="euiTableRowCell__mobileHeader euiTableRowCell--hideForDesktop"
                    >
                      Description
                    </div>
                    <div
                      class="euiTableCellContent euiTableCellContent--overflowingContent"
                    >
                      <span
                        class="euiToolTipAnchor"
                      >
                        <div
                          class="euiText euiText--extraSmall eui-textTruncate"
                          style="max-width: 150px;"
                        >
                          should be able to see the description tooltip when hovering over the description
                        </div>
                      </span>
                    </div>
                  </td>
                  <td
                    class="euiTableRowCell"
                    style="width: 10%;"
                  >
                    <div
                      class="euiTableRowCell__mobileHeader euiTableRowCell--hideForDesktop"
                    >
                      Permissions
                    </div>
                    <div
                      class="euiTableCellContent euiTableCellContent--overflowingContent"
                    >
                      <span
                        class="euiToolTipAnchor"
                      >
                        <div
                          class="euiText euiText--extraSmall"
                        >
                          Admin
                        </div>
                      </span>
                    </div>
                  </td>
                  <td
                    class="euiTableRowCell euiTableRowCell--hasActions"
                  >
                    <div
                      class="euiTableCellContent euiTableCellContent--alignRight euiTableCellContent--showOnHover euiTableCellContent--overflowingContent"
                    >
                      <span
                        class="euiToolTipAnchor"
                      >
                        <button
                          aria-labelledby="generated-id"
                          class="euiButtonIcon euiButtonIcon--primary euiButtonIcon--empty euiButtonIcon--xSmall euiTableCellContent__hoverItem"
                          data-test-subj="workspace-list-copy-id-icon"
                          type="button"
                        >
                          <span
                            aria-hidden="true"
                            class="euiButtonIcon__icon"
                            color="inherit"
                            data-euiicon-type="copy"
                          />
                        </button>
                        <span
                          class="euiScreenReaderOnly"
                          id="generated-id"
                        >
                          <div
                            class="euiText euiText--medium"
                          >
                            Copy ID
                          </div>
                        </span>
                      </span>
                      <span
                        class="euiToolTipAnchor"
                      >
                        <button
                          aria-labelledby="generated-id"
                          class="euiButtonIcon euiButtonIcon--primary euiButtonIcon--empty euiButtonIcon--xSmall euiTableCellContent__hoverItem"
                          data-test-subj="workspace-list-set-default-icon"
                          type="button"
                        >
                          <span
                            aria-hidden="true"
                            class="euiButtonIcon__icon"
                            color="inherit"
                            data-euiicon-type="flag"
                          />
                        </button>
                        <span
                          class="euiScreenReaderOnly"
                          id="generated-id"
                        >
                          <div
                            class="euiText euiText--medium"
                          >
                            Set as my default
                          </div>
                        </span>
                      </span>
                    </div>
                  </td>
                </tr>
                <tr
                  class="euiTableRow euiTableRow-hasActions"
                >
                  <td
                    class="euiTableRowCell"
                    style="width: 25%;"
                  >
                    <div
                      class="euiTableRowCell__mobileHeader euiTableRowCell--hideForDesktop"
                    >
                      Name
                    </div>
                    <div
                      class="euiTableCellContent euiTableCellContent--overflowingContent"
                    >
                      <span
                        class=""
                      >
                        <a
                          class="euiLink euiLink--primary"
                          href="http://localhost/w/id3/undefined"
                          rel="noreferrer"
                        >
                          <div
                            class="euiFlexGroup euiFlexGroup--gutterExtraSmall euiFlexGroup--alignItemsCenter euiFlexGroup--directionRow euiFlexGroup--responsive"
                          >
                            <div
                              class="euiFlexItem"
                            >
                              <div
                                class="euiText euiText--small"
                              >
                                name3
                              </div>
                            </div>
                          </div>
                        </a>
                      </span>
                    </div>
                  </td>
                  <td
                    class="euiTableRowCell"
                    style="width: 20%;"
                  >
                    <div
                      class="euiTableRowCell__mobileHeader euiTableRowCell--hideForDesktop"
                    >
                      Use case
                    </div>
                    <div
                      class="euiTableCellContent"
                    >
                      <span
                        class="euiTableCellContent__text"
                      >
                        Search
                      </span>
                    </div>
                  </td>
                  <td
                    class="euiTableRowCell"
                    style="width: 25%;"
                  >
                    <div
                      class="euiTableRowCell__mobileHeader euiTableRowCell--hideForDesktop"
                    >
                      Description
                    </div>
                    <div
                      class="euiTableCellContent euiTableCellContent--overflowingContent"
                    >
                      <span
                        class="euiToolTipAnchor"
                      >
                        <div
                          class="euiText euiText--extraSmall eui-textTruncate"
                          style="max-width: 150px;"
                        />
                      </span>
                    </div>
                  </td>
                  <td
                    class="euiTableRowCell"
                    style="width: 10%;"
                  >
                    <div
                      class="euiTableRowCell__mobileHeader euiTableRowCell--hideForDesktop"
                    >
                      Permissions
                    </div>
                    <div
                      class="euiTableCellContent euiTableCellContent--overflowingContent"
                    >
                      <span
                        class="euiToolTipAnchor"
                      >
                        <div
                          class="euiText euiText--extraSmall"
                        >
                          Admin
                        </div>
                      </span>
                    </div>
                  </td>
                  <td
                    class="euiTableRowCell euiTableRowCell--hasActions"
                  >
                    <div
                      class="euiTableCellContent euiTableCellContent--alignRight euiTableCellContent--showOnHover euiTableCellContent--overflowingContent"
                    >
                      <span
                        class="euiToolTipAnchor"
                      >
                        <button
                          aria-labelledby="generated-id"
                          class="euiButtonIcon euiButtonIcon--primary euiButtonIcon--empty euiButtonIcon--xSmall euiTableCellContent__hoverItem"
                          data-test-subj="workspace-list-copy-id-icon"
                          type="button"
                        >
                          <span
                            aria-hidden="true"
                            class="euiButtonIcon__icon"
                            color="inherit"
                            data-euiicon-type="copy"
                          />
                        </button>
                        <span
                          class="euiScreenReaderOnly"
                          id="generated-id"
                        >
                          <div
                            class="euiText euiText--medium"
                          >
                            Copy ID
                          </div>
                        </span>
                      </span>
                      <span
                        class="euiToolTipAnchor"
                      >
                        <button
                          aria-labelledby="generated-id"
                          class="euiButtonIcon euiButtonIcon--primary euiButtonIcon--empty euiButtonIcon--xSmall euiTableCellContent__hoverItem"
                          data-test-subj="workspace-list-set-default-icon"
                          type="button"
                        >
                          <span
                            aria-hidden="true"
                            class="euiButtonIcon__icon"
                            color="inherit"
                            data-euiicon-type="flag"
                          />
                        </button>
                        <span
                          class="euiScreenReaderOnly"
                          id="generated-id"
                        >
                          <div
                            class="euiText euiText--medium"
                          >
                            Set as my default
                          </div>
                        </span>
                      </span>
                    </div>
                  </td>
                </tr>
              </tbody>
            </table>
          </div>
          <div>
            <div
              class="euiSpacer euiSpacer--m"
            />
            <div
              class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow"
            >
              <div
                class="euiFlexItem euiFlexItem--flexGrowZero"
              >
                <div
                  class="euiPopover euiPopover--anchorUpRight"
                >
                  <div
                    class="euiPopover__anchor"
                  >
                    <button
                      class="euiButtonEmpty euiButtonEmpty--text euiButtonEmpty--xSmall"
                      data-test-subj="tablePaginationPopoverButton"
                      type="button"
                    >
                      <span
                        class="euiButtonContent euiButtonContent--iconRight euiButtonEmpty__content"
                      >
                        <span
                          class="euiButtonContent__icon"
                          color="inherit"
                          data-euiicon-type="arrowDown"
                        />
                        <span
                          class="euiButtonEmpty__text"
                        >
                          Rows per page
                          : 
                          10
                        </span>
                      </span>
                    </button>
                  </div>
                </div>
              </div>
              <div
                class="euiFlexItem euiFlexItem--flexGrowZero"
              >
                <nav
                  class="euiPagination"
                >
                  <button
                    aria-label="Previous page"
                    class="euiButtonIcon euiButtonIcon--text euiButtonIcon--empty euiButtonIcon--xSmall"
                    data-test-subj="pagination-button-previous"
                    disabled=""
                    type="button"
                  >
                    <span
                      aria-hidden="true"
                      class="euiButtonIcon__icon"
                      color="inherit"
                      data-euiicon-type="arrowLeft"
                    />
                  </button>
                  <ul
                    class="euiPagination__list"
                  >
                    <li
                      class="euiPagination__item"
                    >
                      <button
                        aria-controls="__table_generated-id"
                        aria-current="true"
                        aria-label="Page 1 of 1"
                        class="euiButtonEmpty euiButtonEmpty--text euiButtonEmpty--small euiButtonEmpty-isDisabled euiPaginationButton euiPaginationButton-isActive euiPaginationButton--hideOnMobile"
                        data-test-subj="pagination-button-0"
                        disabled=""
                        type="button"
                      >
                        <span
                          class="euiButtonContent euiButtonEmpty__content"
                        >
                          <span
                            class="euiButtonEmpty__text"
                          >
                            1
                          </span>
                        </span>
                      </button>
                    </li>
                  </ul>
                  <button
                    aria-label="Next page"
                    class="euiButtonIcon euiButtonIcon--text euiButtonIcon--empty euiButtonIcon--xSmall"
                    data-test-subj="pagination-button-next"
                    disabled=""
                    type="button"
                  >
                    <span
                      aria-hidden="true"
                      class="euiButtonIcon__icon"
                      color="inherit"
                      data-euiicon-type="arrowRight"
                    />
                  </button>
                </nav>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
